### PR TITLE
[AMDGPU] Fix mode register pass for constrained FP operations

### PR DESCRIPTION
**Pull request by abhigargrepo**
This PR will fix the si-mode-register pass which is inserting an extra setreg instruction in case of constrained FP operations. This pass will be ignored for strictfp functions.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll [hunk 1]
```diff
@@ -9,8 +9,7 @@ define double @ignoreStrictfp(double noundef %a, double noundef %b) #0 {
 ; GCN:       ; %bb.0:
 ; GCN-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
 ; GCN-NEXT:    s_setreg_imm32_b32 hwreg(HW_REG_MODE, 2, 2), 1
-; GCN-NEXT:    s_nop 1
-; GCN-NEXT:    s_setreg_imm32_b32 hwreg(HW_REG_MODE, 2, 1), 0
+; GCN-NOT:     s_setreg_imm32_b32 hwreg(HW_REG_MODE, 2, 1), 0
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll:12]: Don't manually add the -NOT here. The generated -NEXT checks are sufficient and much more reliable

File: llvm/lib/Target/AMDGPU/SIModeRegister.cpp [hunk 2]
```diff
@@ -430,6 +430,9 @@ void SIModeRegister::processBlockPhase3(MachineBasicBlock &MBB,
 }
 
 bool SIModeRegister::runOnMachineFunction(MachineFunction &MF) {
+  const Function &F = MF.getFunction();
+  if (F.hasFnAttribute(llvm::Attribute::StrictFP))
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIModeRegister.cpp:434]: Add a comment explaining why

Comment (abhigargrepo) [reply llvm/lib/Target/AMDGPU/SIModeRegister.cpp:434]: Will update

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIModeRegister.cpp:438]: This comment doesn't really explain why though. Should say something about how strictfp functions have user managed FP modes

Comment (abhigargrepo) [reply llvm/lib/Target/AMDGPU/SIModeRegister.cpp:439]: Updated, kindly review

**Review by arsenm — APPROVED** [review]

**Pull request merged**